server/plugin_release.go

```diff
@@ -29,199 +35,406 @@ import (
-	LogInfo("Getting the asset download URL")
-	releaseAsset, _, err := client.Repositories.GetReleaseAsset(ctx, Cfg.GithubOrg, repo, assetID)
+// verifySignatures verifies plugin files, assumes signatures are <filepath>.sig.
+func verifySignatures(pluginFilePaths []string) error {
+	block, err := armor.Decode(bytes.NewReader(mattermostPluginPublicKey))
 	if err != nil {
-		return "", err
+		return fmt.Errorf("failed to decode public key err=%w", err)
+	}
+
+	keyring, err := openpgp.ReadKeyRing(block.Body)
+	if err != nil {
+		return fmt.Errorf("can't read public key err=%w", err)
+	}
+
+	for _, pluginFilePath := range pluginFilePaths {
+		signedFile, err := os.Open(pluginFilePath)
+		if err != nil {
+			return fmt.Errorf("cannot read signed file err=%w", err)
+		}
+		defer signedFile.Close()
+
+		// Assume signature is always <filepath>.sig
+		signatureFile, err := os.Open(fmt.Sprintf("%s.sig", pluginFilePath))
+		if err != nil {
+			return fmt.Errorf("cannot read signature file err=%w", err)
+		}
+		defer signatureFile.Close()
+
+		if _, err = openpgp.CheckDetachedSignature(keyring, signedFile, signatureFile); err != nil {
+			return fmt.Errorf("error while checking the signature err=%w", err)
+		}
 	}
 
-	return releaseAsset.GetBrowserDownloadURL(), nil
+	LogInfo("Signatures verified for %+v", pluginFilePaths)
+	return nil
 }
 
-func uploadSignedArtifcatsToGithub(ctx context.Context, githubClient *github.Client, repo string, releaseID int64, tempFolder string, fileToUpload []string) error {
-	LogInfo("Uploading signed assets to Github release")
+// createPlatformPlugins splits plugin tar into platform specific plugin tars.
+// Returns paths to platform plugin tars if successful, or an error otherwise.
+func createPlatformPlugins(repositoryName, tag, pluginFilePath, pluginFolder string) ([]string, error) {
+	var result []string
+	if err := hasAllPlatformBinaries(pluginFilePath); err != nil {
+		return nil, err
+	}
 
-	for _, file := range fileToUpload {
-		opts := &github.UploadOptions{
-			Name: file,
+	platformFileExclusion := map[string][]string{
+		"osx-amd64":     {"windows", "linux"},
+		"windows-amd64": {"darwin", "linux"},
+		"linux-amd64":   {"darwin", "windows"},
+	}
+
+	platformFileInclusion := map[string]string{
+		"osx-amd64":     "plugin-darwin-amd64",
+		"windows-amd64": "plugin-windows-amd64.exe",
+		"linux-amd64":   "plugin-linux-amd64",
+	}
+
+	for platform, excludeOtherPlatforms := range platformFileExclusion {
+		deleteWildcards := ""
+		for _, deleteWildcard := range excludeOtherPlatforms {
+			deleteWildcards += fmt.Sprintf(`--delete "*%s*" `, deleteWildcard)
 		}
 
-		filePath := fmt.Sprintf("/%s/%s", tempFolder, file)
-		fileHandler, err := os.Open(filePath)
-		defer fileHandler.Close()
-		if err != nil {
-			LogError("Error opening the file. err=" + err.Error())
-			return err
+		// Couldn't achieve gzip level compressions with golang archive api, using shell cmds instead.
```

Comment:
agreed, however i couldn't do without using the bash commands here.

The issue I was was the bash command used gzip which uses `LZ77` compression when gzipping, however there is no `LZ77` encoder/decoder in the [std go lib](https://stackoverflow.com/a/28252723/5551762) unfortunately and the archive api wasn't giving me as good of a compression as i was getting with these command, hence i decided to stick with them.

The difference was about a few MBs, but since we're prepackaging these files, i thought it'd be better if we can get the best compressions.

Reply:
Thanks for the explanation. How many MBs are saved with using `exec`? I'm wondering if the cost is worth the risk.

Reply:
i think the difference was around 3-4MBs not huge, however since any changes to this files affects gitlab binaries, i didn't want to make a change that affected them without giving them a headsup. As things stand, the sizes are the same so it's not affecting them.